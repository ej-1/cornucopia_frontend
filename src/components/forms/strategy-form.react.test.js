import React from "react";
import StrategyForm from "../forms/strategy-form";
import renderer from "react-test-renderer";
import { shallow, mount } from "enzyme";
import moment from "moment";

const fetchSimulation = jest.fn();

const form = shallow(<StrategyForm runSimulation={fetchSimulation} />);

const component = renderer.create(<StrategyForm />);

// Test that handleSubmit() calls parent's function.
// Mock testing function call with arguments
// https://stackoverflow.com/questions/43245040/using-jest-to-spy-on-method-call-in-componentdidmount
test("handleSubmit, sends correct data when state is set", () => {
  const wrapper = mount(<StrategyForm runSimulation={fetchSimulation} />);
  let data = {
    amount: "1000",
    currencyPair: "BTC-USDT",
    startDate: moment("2010-01-01"),
    endDate: moment("2010-12-31"),
    strategy: "MACD"
  };
  wrapper.setState(data);

  wrapper.instance().handleSubmit();

  expect(fetchSimulation).toHaveBeenCalledWith(data);
});

// Test that snapshot sets state attributes on clicking form submit button.
test("User fills in form and submit, setting state fields", () => {
  const wrapper = mount(<StrategyForm runSimulation={fetchSimulation} />);
  //snap.toMatchSpecificSnapshot('./specific/dir/my.shot');
  let data = {
    amount: "1000",
    currencyPair: "BTC-USDT",
    startDate: moment("2010-01-01"),
    endDate: moment("2010-12-31"),
    strategy: "MACD"
  };

  const dropdownCurrency = wrapper.find("#dropdown-currency");
  dropdownCurrency
    .find("ul .dropdown-menu li")
    .at(0)
    .find("a")
    .simulate("click");

  const dropdownStrategy = wrapper.find("#dropdown-strategy");
  dropdownStrategy
    .find("ul .dropdown-menu li")
    .at(0)
    .find("a")
    .simulate("click");

  const datepickerStartdate = wrapper.find("#datepicker-startdate");
  datepickerStartdate
    .hostNodes()
    .simulate("change", { target: { value: moment("2012-01-01") } });

  const datepickerEnddate = wrapper.find("#datepicker-enddate");
  datepickerEnddate;
  wrapper
    .find("#datepicker-enddate")
    .hostNodes()
    .simulate("change", { target: { value: moment("2013-01-01") } });

  wrapper.find("form").simulate("submit", data);

  expect(wrapper.state()).toEqual({
    currencyPair: "BTC-USDT",
    strategy: "MACD",
    startDate: moment("2012-01-01"),
    endDate: moment("2013-01-01")
  });

  expect(fetchSimulation).toHaveBeenCalledWith(data);
});

// DON'T NEED TO TEST
// that dropdown works as intended. Trust in react-bootstrap.

// test datepicker that it changes startDate and endDate.

// test input that it changes startDate and endDate.

/*
test("Dropdown selected values changes on select", () => {
  let dropdown = form.find("#dropdown-currencypair");
  let menuitem = dropdown.childAt(0);
  expect(menuitem.childAt(0).text()).toEqual("BTC-BNB");
  menuitem.simulate("click");
  //expect(form.props()).toEqual("blaasaa");
  //expect(form.find("#dropdown-currencypair").html()).toEqual("bla");
  let tree = component.toJSON();
  tree.props.handleChangeCurrencyPair("BNB-USDT");
  tree = component.toJSON();
  expect(tree).toEqual("va");
  //expect(menuitem.html()).toEqual(""<li role=\"presentation\" class=\"\"><a role=\"menuitem\" tabindex=\"-1\" href=\"#\">BTC-BNB</a></li>"");
});

test("Link changes the class when hovered", () => {
  const component = renderer.create(<StrategyForm />);
  let tree = component.toJSON();
  //expect(tree).toMatchSnapshot();

  // manually trigger the callback
  //tree.props.handleChangeCurrencyPair("BNB-USDT");
  // re-rendering
  tree = component.toJSON();
  console.log(tree.children[0].children[0]);
  //expect(tree.children).toMatchSnapshot();
});


test("Dropdown defautl value", () => {
  let dropdown = form.find("#dropdown-currencypair");
  expect(dropdown.html()).toEqual(
    '<div class="dropdown btn-group"><button id="dropdown-currencypair" role="button" aria-haspopup="true" aria-expanded="false" type="button" class="dropdown dropdown-toggle btn btn-default">Currency Pair <span class="caret"></span></button><ul role="menu" class="dropdown-menu" aria-labelledby="dropdown-currencypair"><li role="presentation" class=""><a role="menuitem" tabindex="-1" href="#">BTC-BNB</a></li><li role="presentation" class=""><a role="menuitem" tabindex="-1" href="#">BTC-USDT</a></li></ul></div>'
  );
});
*/
/*
  const component = renderer.create(<StrategyForm />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
*/

// if IMPORT SYNTAX ERROR - solution : yarn add --dev babel-jest babel-core regenerator-runtime
//
// if TypeError: environment.teardown is not a function - solution: npm install jest-cli
//
//
