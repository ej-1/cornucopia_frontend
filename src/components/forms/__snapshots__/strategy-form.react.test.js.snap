// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown defautl value 1`] = `
<div
  className="strategy-form-container"
>
  <div
    className="strategy-form"
  >
    <h4>
      Simulate strategy
    </h4>
    <form
      id="strategy-form"
      name="strategy-form"
      onSubmit={[Function]}
    >
      <label
        className="control-label"
        htmlFor={undefined}
      >
        Amount
      </label>
       
      <input
        className="strategy-form-amount form-control"
        id={undefined}
        onChange={[Function]}
        placeholder={undefined}
        type="number"
        value={1000}
      />
      <div
        className="dropdown btn-group"
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="dropdown dropdown-toggle btn btn-default"
          disabled={false}
          id="dropdown-currencypair"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          type="button"
        >
          Currency Pair
           
          <span
            className="caret"
          />
        </button>
        <ul
          aria-labelledby="dropdown-currencypair"
          className="dropdown-menu"
          role="menu"
        >
          <li
            className=""
            role="presentation"
            style={undefined}
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              BTC-BNB
            </a>
          </li>
          <li
            className=""
            role="presentation"
            style={undefined}
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              BTC-USDT
            </a>
          </li>
        </ul>
      </div>
      <div
        className="dropdown btn-group"
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="dropdown dropdown-toggle btn btn-default"
          disabled={false}
          id="dropdown-strategy"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          type="button"
        >
          Strategy
           
          <span
            className="caret"
          />
        </button>
        <ul
          aria-labelledby="dropdown-strategy"
          className="dropdown-menu"
          role="menu"
        >
          <li
            className=""
            role="presentation"
            style={undefined}
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              MACD
            </a>
          </li>
          <li
            className=""
            role="presentation"
            style={undefined}
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              EMA
            </a>
          </li>
        </ul>
      </div>
      <div>
        <label
          className="control-label"
          htmlFor={undefined}
        >
          Trading Start date
        </label>
         
        <div>
          <div
            className="react-datepicker-wrapper"
          >
            <div
              className="react-datepicker__input-container"
            >
              <input
                autoComplete={undefined}
                autoFocus={undefined}
                className=""
                disabled={false}
                id={undefined}
                name={undefined}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder={undefined}
                readOnly={undefined}
                required={undefined}
                tabIndex={undefined}
                title={undefined}
                type="text"
                value="2018/05/28 19:00"
              />
            </div>
          </div>
        </div>
        <label
          className="control-label"
          htmlFor={undefined}
        >
          Trading End date
        </label>
         
        <div>
          <div
            className="react-datepicker-wrapper"
          >
            <div
              className="react-datepicker__input-container"
            >
              <input
                autoComplete={undefined}
                autoFocus={undefined}
                className=""
                disabled={false}
                id={undefined}
                name={undefined}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder={undefined}
                readOnly={undefined}
                required={undefined}
                tabIndex={undefined}
                title={undefined}
                type="text"
                value="2018/05/28 19:00"
              />
            </div>
          </div>
        </div>
      </div>
      <button
        className="btn btn-warning btn-block"
        disabled={false}
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;

exports[`Link changes the class when hovered 1`] = `
<div
  className="strategy-form-container"
>
  <div
    className="strategy-form"
  >
    <h4>
      Simulate strategy
    </h4>
    <form
      id="strategy-form"
      name="strategy-form"
      onSubmit={[Function]}
    >
      <label
        className="control-label"
        htmlFor={undefined}
      >
        Amount
      </label>
       
      <input
        className="strategy-form-amount form-control"
        id={undefined}
        onChange={[Function]}
        placeholder={undefined}
        type="number"
        value={1000}
      />
      <div
        className="dropdown btn-group"
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="dropdown-currencypair dropdown-toggle btn btn-default"
          disabled={false}
          id="dropdown-currencypair"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          type="button"
        >
          Currency Pair
           
          <span
            className="caret"
          />
        </button>
        <ul
          aria-labelledby="dropdown-currencypair"
          className="dropdown-menu"
          role="menu"
        >
          <li
            className=""
            role="presentation"
            style={undefined}
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              BTC-BNB
            </a>
          </li>
          <li
            className=""
            role="presentation"
            style={undefined}
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              BTC-USDT
            </a>
          </li>
        </ul>
      </div>
      <div
        className="dropdown btn-group"
      >
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="dropdown dropdown-toggle btn btn-default"
          disabled={false}
          id="dropdown-strategy"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          type="button"
        >
          Strategy
           
          <span
            className="caret"
          />
        </button>
        <ul
          aria-labelledby="dropdown-strategy"
          className="dropdown-menu"
          role="menu"
        >
          <li
            className=""
            role="presentation"
            style={undefined}
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              MACD
            </a>
          </li>
          <li
            className=""
            role="presentation"
            style={undefined}
          >
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1"
            >
              EMA
            </a>
          </li>
        </ul>
      </div>
      <div>
        <label
          className="control-label"
          htmlFor={undefined}
        >
          Trading Start date
        </label>
         
        <div>
          <div
            className="react-datepicker-wrapper"
          >
            <div
              className="react-datepicker__input-container"
            >
              <input
                autoComplete={undefined}
                autoFocus={undefined}
                className=""
                disabled={false}
                id={undefined}
                name={undefined}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder={undefined}
                readOnly={undefined}
                required={undefined}
                tabIndex={undefined}
                title={undefined}
                type="text"
                value="2018/05/28 22:00"
              />
            </div>
          </div>
        </div>
        <label
          className="control-label"
          htmlFor={undefined}
        >
          Trading End date
        </label>
         
        <div>
          <div
            className="react-datepicker-wrapper"
          >
            <div
              className="react-datepicker__input-container"
            >
              <input
                autoComplete={undefined}
                autoFocus={undefined}
                className=""
                disabled={false}
                id={undefined}
                name={undefined}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder={undefined}
                readOnly={undefined}
                required={undefined}
                tabIndex={undefined}
                title={undefined}
                type="text"
                value="2018/05/28 22:00"
              />
            </div>
          </div>
        </div>
      </div>
      <button
        className="btn btn-warning btn-block"
        disabled={false}
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
